### DISCUSSION TITLE
DotNet.Testcontainers.Containers.ResourceReaperException : Initialization has been cancelled.

### DISCUSSION THREAD

**Discussion by dpraimeyuu** [testcontainers/testcontainers-dotnet#443] [Q&A] [answered]
Hi @HofmeisterAn!

First and foremost, thanks for your this package and great effort in maintaining it! ❤️ 

I wanted to use `dotnet-testcontainers` in e2e tests with combination of `WebApplicationFactory` from `net5.0` so I created the following extension method:
```cs
public static async Task<IServiceProvider> StartSqlServerTestContainerAsync(this IServiceProvider serviceProvider)
{
    var maintenanceContext = serviceProvider.GetRequiredService<MaintenanceContext>();
    string connectionString = maintenanceContext.Database.GetConnectionString();
    var dbConnectionStringBuilder = new SqlConnectionStringBuilder(connectionString);
    int portNumber = int.Parse(dbConnectionStringBuilder.DataSource.Split(",")[1]);

    ITestcontainersBuilder<MsSqlTestcontainer> testContainersBuilder
        = new TestcontainersBuilder<MsSqlTestcontainer>()
                .WithDatabase(new MsSqlTestcontainerConfiguration
                {
                    Password = dbConnectionStringBuilder.Password,
                    Port = portNumber
                })
                .WithName("maintenance")
                .WithImage("mcr.microsoft.com/mssql/server:2019-CU15-ubuntu-20.04");

    MsSqlTestcontainer dbTestContainer = testContainersBuilder.Build();
    try
    {
        await dbTestContainer.StartAsync();
    }
    catch (Exception)
    {
        throw;
    }

    return serviceProvider;
}
```
Then used in the following method:
```cs
public async Task Setup()
{
    await Services.StartSqlServerTestContainerAsync();
    Services.RunEfCoreMigrations();
}
```
When running `Setup` in the test method, after a period of time such exception is thrown:
`
DotNet.Testcontainers.Containers.ResourceReaperException : Initialization has been cancelled.
    Stack Trace:
       at DotNet.Testcontainers.Containers.ResourceReaper.GetAndStartNewAsync(Guid sessionId, String ryukImage, TimeSpan initTimeout, CancellationToken ct)
   at DotNet.Testcontainers.Containers.ResourceReaper.GetAndStartNewAsync(Guid sessionId, String ryukImage, TimeSpan initTimeout, CancellationToken ct)
   at DotNet.Testcontainers.Containers.ResourceReaper.GetAndStartDefaultAsync(CancellationToken ct)
   at DotNet.Testcontainers.Clients.TestcontainersClient.RunAsync(ITestcontainersConfiguration configuration, CancellationToken ct)
   at DotNet.Testcontainers.Containers.TestcontainersContainer.Create(CancellationToken ct)
   at DotNet.Testcontainers.Containers.TestcontainersContainer.StartAsync(CancellationToken ct)
`
From time to time I see tests running properly and two containers are available in the Docker Desktop panel, but in most cases only one is visible:
![image](https://user-images.githubusercontent.com/6437191/162899967-8672be1a-2718-4916-ad5f-206b5ae69467.png)

My setup is: Docker Desktop, Linux containers, WSL2.

I assume that when a container with `MSSQL` tries to be created something "bad" happens, but I don't know how to inspect it so that I am able to provide some more insights.

What's more interesting, when there's no image of `MSSQL` available locally - so docker pulls it for the first time - **then tests pass with success.**

I wonder about the possible cause of such a scenario?

**Comment by HofmeisterAn** [comment, marked as answer]
It seems like you're having an issue I saw a few times in the past too. Until now, I couldn't reproduce it. Maybe we can figure it out together. Do you use the latest beta release?

> When running `Setup` in the test method, after a period of time such exception is thrown:
How much time is in between?

You'll get this exception if Testcontainers can't establish a connection to the Resource Reaper (Ryuk) within 10 seconds.
https://github.com/HofmeisterAn/dotnet-testcontainers/blob/b99957969f1bf8d1615196f1a77899cb5d51c76e/src/DotNet.Testcontainers/Containers/ResourceReaper.cs#L229

Testcontainers starts with the first container, the Resource Reaper, in the background. Can you start the Resource Reaper in advance:

```cs
await ResourceReaper.GetAndStartDefaultAsync()
  .ConfigureAwait(false);
```

to check if the connection works at all?

This may cancel your test:

https://github.com/HofmeisterAn/dotnet-testcontainers/blob/b99957969f1bf8d1615196f1a77899cb5d51c76e/src/DotNet.Testcontainers/Containers/ResourceReaper.cs#L156

**Reply by dpraimeyuu** [reply]
Thanks for replying!

I started `ResourceReaper` in advance, as you suggested. Here's my `Setup` method:
```cs
public async Task Setup()
{
    await ResourceReaper.GetAndStartDefaultAsync()
        .ConfigureAwait(false);
    await Services.StartSqlServerTestContainerAsync();
    Services.RunEfCoreMigrations();
}
```
The same effect: `ResourceReaperException` with the same message.

Additionally, I checked logs of `ryuk` container and I only see:
![image](https://user-images.githubusercontent.com/6437191/162923212-39cb381f-4eac-4c26-af68-a31cd9ec7e81.png)

EDIT:
Another trial and test execution resulted in:
![image](https://user-images.githubusercontent.com/6437191/162923453-3e3094fa-9832-449b-b888-50199fbad3c3.png)

**Reply by HofmeisterAn** [reply]
For some reason, Testcontainers can't connect to the Ryuk in the first run / screenshot. Any chances you can enable logging or add some debug messages? Does it fail immediately? Especially, the content of `messageBuffer` would be interessting.

**Reply by dpraimeyuu** [reply]
@HofmeisterAn would this:
```cs
TestcontainersSettings.Logger = new Logger<MsSqlTestcontainer>(new LoggerFactory());
```
be enough?

Where I can find `messageBuffer`? Is it some property of `ResourceReaper`?

**Reply by HofmeisterAn** [reply]
> be enough?

Yes, as long as it implements `Microsoft.Extensions.Logging.ILogger`. You can use the built-in or [third-party](https://docs.microsoft.com/dotnet/core/extensions/logging-providers#third-party-logging-providers) logging providers.

> Where I can find `messageBuffer`? Is it some property of `ResourceReaper`?

Yes, it is:

https://github.com/HofmeisterAn/dotnet-testcontainers/blob/b99957969f1bf8d1615196f1a77899cb5d51c76e/src/DotNet.Testcontainers/Containers/ResourceReaper.cs#L284

**Reply by HofmeisterAn** [reply]
The latest version (`1.6.0-beta.2185`) contains some more Resource Reaper improvements. Even I think they are not related to your issue, you can at least try that version. If it still does not work, I need the debug logs.

**Reply by srollinet** [reply]
@dpraimeyuu I had issues on windows + WSL2 with another test container lib (https://github.com/isen-ng/testcontainers-dotnet/issues/58)
The port that was auto-assigned by Docker for Ryuk was reserved in Windows. Therefore it was not possible to connect to Ryuk.

Do you think it can be related?

**Reply by dpraimeyuu** [reply]
@srollinet thanks for your reply! Might be, I haven't checked that. Then a natural question comes to my mind whether a specific port declaration will solve the issue? Did you manage to try it?

**Reply by srollinet** [reply]
@dpraimeyuu Yes, using a specific port solved the problem. I don't know if the option exists in dotnet-testcontainers.
For the other lib, I authored a PR to allow setting the Ryuk port

**Reply by HofmeisterAn** [reply]
@srollinet Interesting, I was not aware of this issue. Unfortunately, using a specific port (default configuration) might be a problem for others. However, we could support specific ports with `GetAndStartNewAsync`. In that case, the developer has to configure the resource reaper and pass its session ID to each resource (instead of using the default configuration).

We need to extend `GetAndStartNewAsync` then:
https://github.com/HofmeisterAn/dotnet-testcontainers/blob/ba7d87befb0ece496d94ec9f350d503a264a372c/src/DotNet.Testcontainers/Containers/ResourceReaper.cs#L123

**Reply by srollinet** [reply]
@HofmeisterAn  In testcontainers-dotnet, it is possible to define an env variable with the Ryuk port. It might also be a possibility, or to define a static variable.
But this kind of solution doesn't seem very clean and may be quite confusing (It must be set before the first container is started, otherwise it has no effect)

Another solution is maybe to find an open port using c# code, so we ensure the port is open in Windows.

**Reply by davidpetric** [reply]
Had the same issue, updating the Testcontainers NuGet package from 2.1.0 to 2.2.0-beta solved the issue for me.

Thanks!
